### PR TITLE
feat: implement async internal package

### PR DESCRIPTION
**Pull request by QuentinBrosse**

### CODE REVIEW

File: internal/async/wait.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	"time"
+)
+
+var (
+	defaultInterval = time.Second
+	defaultTimeout  = time.Minute * 5
+)
+
+// WaitOptions defines the waiting options.
+type WaitSyncConfig struct {
```
Comment (kindermoumoute) [inline internal/async/wait.go]: WaitSyncRequest?

File: internal/async/wait.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	"fmt"
+	"time"
+)
+
+var (
+	defaultInterval = time.Second
+	defaultTimeout  = time.Minute * 5
+)
+
+// WaitOptions defines the waiting options.
```
Comment (kindermoumoute) [inline internal/async/wait.go]: Typo

File: internal/async/wait.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+type WaitSyncConfig struct {
+	get      func() (value interface{}, err error, stopCondition bool)
+	interval func() <-chan time.Time
```
Comment (kindermoumoute) [inline internal/async/wait.go]: rename to intervalFunc

File: internal/async/wait.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	if config.timeout == 0 {
+		config.timeout = defaultTimeout
+	}
+
+	type payload struct {
+		value interface{}
+		err   error
+	}
+
+	result := make(chan payload, 1)
```
Comment (kindermoumoute) [inline internal/async/wait.go]: remove `,1`

File: internal/async/wait.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	timeout := make(chan bool, 1)
```
Comment (kindermoumoute) [inline internal/async/wait.go]: remove `, 1`

File: internal/async/wait.go (hunk 3 above)
Comment (jerome-quere) [inline internal/async/wait.go]:
```
type intervalStrategy func() <-chan time.Time
intervalStrategy intervalStrategy
```

File: internal/async/wait.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import (
+	"fmt"
+	"time"
+)
+
+var (
+	defaultInterval = time.Second
+	defaultTimeout  = time.Minute * 5
+)
+
```
Comment (jerome-quere) [inline internal/async/wait.go]:
```
func LinearIntervalStrategy(interval time.Duration) intervalStrategy {
  return  func() <-chan time.Time {
			return time.After(interval)
  }
}

func FibonacciIntervalStrategy(base time.Duration, factor double = 1) intervalStrategy {
  return  func() <-chan time.Time {
  }
}
```

File: internal/async/wait.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+	timeout  time.Duration
+}
+
+// WaitSync waits and returns when a given stop condition is true or if an error occurs.
+//
+// In this sync implementation, the timeout could be longer than expected.
+func WaitSync(config *WaitSyncConfig) (terminalValue interface{}, err error) {
+	// initialize configuration
+	if config.interval == nil {
+		config.interval = func() <-chan time.Time {
```
Comment (jerome-quere) [inline internal/async/wait.go]:
```
config.interval = LinearIntervalStrategy(defaultInterval)
```

File: internal/async/wait.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+)
+
+var (
+	defaultInterval = time.Second
+	defaultTimeout  = time.Minute * 5
+)
+
+// WaitOptions defines the waiting options.
+type WaitSyncConfig struct {
+	get      func() (value interface{}, err error, stopCondition bool)
```
Comment (jerome-quere) [inline internal/async/wait.go]: add comment warning
```
// This method will be call from another goroutine
```

Comment (jerome-quere) [inline internal/async/wait.go]: stopCondition => isTerminal

File: internal/async/wait.go [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+	interval func() <-chan time.Time
+	timeout  time.Duration
+}
+
+// WaitSync waits and returns when a given stop condition is true or if an error occurs.
+//
+// In this sync implementation, the timeout could be longer than expected.
```
Comment (jerome-quere) [inline internal/async/wait.go]: remove

File: internal/async/wait.go [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+	if config.interval == nil {
+		config.interval = func() <-chan time.Time {
+			return time.After(defaultInterval)
+		}
+	}
+	if config.timeout == 0 {
+		config.timeout = defaultTimeout
+	}
+
+	type payload struct {
```
Comment (jerome-quere) [inline internal/async/wait.go]: remove and create a 3rd channel
```
	err := make(chan error, 1)
```

File: internal/async/wait_test.go [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+		v := &value{
+			doneIterations: iterations - cpt,
+			totalDuration:  time.Since(startTime),
+		}
+		return v, nil, cpt == 0
+	}
+}
+
+func getFibonacciInterval() func() <-chan time.Time {
```
Comment (jerome-quere) [inline internal/async/wait_test.go]: remove

File: internal/async/wait.go (hunk 1 above)
Comment (QuentinBrosse) [reply internal/async/wait.go]: `WaitSync` very abstracted and could be used to other use cases than waiting for a request. Therefore, I think that the concept of "config" is better than "request".

**Pull request merged**